File: 121_best_time_to_buy_and_sell_stock/maxProfit.go
func maxProfit(prices []int) int {
    l, r, max := 0, 1, 0

    for r < len(prices) {
        candidate := prices[r] - prices[l]
        if prices[r] > prices[l] {
            max = findMax(candidate, max)
        } else {
            l = r
        }

        r++
    }


    return max
}

func findMax(a int, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

File: 121_best_time_to_buy_and_sell_stock/maxProfit.java
class Solution {
    public int maxProfit(int[] prices) {
        int left = 0;
        int right = 1;
        int max = 0;

        while (right < prices.length) {
            int candidate = prices[right] - prices[left];
            if (prices[right] > prices[left]) {
                max = Math.max(candidate, max);
            } else {
                left = right;
            }
            right++;
        }

        return max;
    }
}